Hesa support inner packages

hesa opened the pull request:
Adding support to read (and use) inner packages (packages within the same SBoM) in relationship. If relationship is a dependency than the package is added as a dependency and used when checking for compatibility.

hesa committed:
Example: add SBoM with inner packages

hesa committed:
Test: add test of inner relations

hesa committed:
Project: Add support for inner packages

Read inner packages and add as dependencies

DEPENDENCY_TAGS - SPDX tags denoting a dependency

hesa committed:
Project: lint issues

jens-erdmann commented on flict/flictlib/project/reader.py:
```diff
@@ -17,7 +19,11 @@ def __init__(self, project_file, project_dirs):
+        for dep in package.get('dependencies', []):
+            if 'license' not in dep:
+                raise FlictError(ReturnCodes.RET_INVALID_PROJECT, f'Dependency does not contain "license": {dep}')
+            licenses.append(f" ( {dep['license']}) ")
```
Why is this reverted back to its old state?

jens-erdmann replied:
Just to elaborate: The former code does exactly the same as the new one with the only difference being the error message starting with `Dependency` instead of `Package`.

hesa replied:
new state back with the coming push

priv-kweihmann commented on example-data/freetype-inner-2.9.spdx.json:
```diff
@@ -0,0 +1,80 @@
+    ]
+  },
+  "name": "freetype-2.9",
+  "dataLicense": "CC0-1.0",
```
This contradicts the GPL-3.0 labeling in the dep5 of reuse

priv-kweihmann replied:
This is still an issue

hesa replied:
Fixed incoming push

```
Files: example-data/osadl-matrix-mini.csv
Copyright: 2021 Henrik Sandklef <hesa@sandklef.com>
License: GPL-3.0-or-later

Files: example-data/*.json
Copyright: 2022 Henrik Sandklef <hesa@sandklef.com>
License: CC0-1.0
```

priv-kweihmann commented on flict/flictlib/project/reader.py:
```diff
@@ -8,6 +8,8 @@
 
 from flict.flictlib.return_codes import FlictError, ReturnCodes
 
+DEPENDENCY_TAGS = ["DYNAMIC_LINK", "STATIC_LINK", "DEPENDS_ON", "CONTAINS", "COPY_OF"]
```
Please refrain from using double quotes - the current style is single ticks only

priv-kweihmann replied:
Still open

hesa replied:
Fixed in coming push

priv-kweihmann commented on flict/flictlib/project/reader.py:
```diff
@@ -140,23 +146,30 @@ def read_project(self, project_file):
-                'version': pkg['versionInfo'],
-                'license': pkg['licenseConcluded'],
-                'description': pkg['description'],
+                'name': pkg.get('name', ""),
```
Same here - inconsistent quote style - please use single ticks only

priv-kweihmann replied:
Additionally this strips the implied error handling - as the name from the package information can now be empty, which is wrong IMO - so the old style at least would have thrown an error

hesa replied:
reverted back away from get
quotes fixed

... in coming push

priv-kweihmann commented on flict/flictlib/project/reader.py:
```diff
@@ -140,23 +146,30 @@ def read_project(self, project_file):
+            top_package = dep['relatedSpdxElement']
+            relationshiptype = dep['relationshipType']
+            if not self._relationship_is_dependency(relationshiptype):
+                logging.debug(f"relationship {dep.get('spdxElementId')} ignored since {relationshiptype} is not defined to be a dependency tag.")
```
AFAIS this code path is missing a test case - please add one

hesa replied:
`    def test_zlib_with_deps(self)` in `tests/test_verify.py`does this

hesa replied:
Added test case as a response to your comment. coming with the push

priv-kweihmann commented on flict/flictlib/project/reader.py:
```diff
@@ -170,6 +183,22 @@ def _read_spdx_2_2(self, only_packages=None):
                     if dep_package_name == _pkg_name:
                         packages[top_package]['dependencies'].append({_pkg_name: _pkg})
+            else:
+                dep_id = dep['spdxElementId']
```
Are you sure about this? I would have expected that you only use the remainder after the `:` for looking up the package.
Also I'm missing a test case for both scenarios (package found and package not found).

hesa replied:
There should be no ":" in dep['spdxElementId'] in this branch so we should be able to use as is. Or?

hesa replied:
Should a missing package (one package refers to another package which is not defined) be raising an error instead? 

My vote: yes

priv-kweihmann replied:
yes. please

hesa replied:
Keep warning for now.
Adding an issue later.

priv-kweihmann commented on flict/flictlib/project/reader.py:
```diff
@@ -140,23 +146,30 @@ def read_project(self, project_file):
+            if not self._relationship_is_dependency(relationshiptype):
+                logging.debug(f"relationship {dep.get('spdxElementId')} ignored since {relationshiptype} is not defined to be a dependency tag.")
+                continue
+            if ":" in dep['spdxElementId']:
```
IMO it would be better to look for `DocumentRef-` prefixes - just the `:` might not be enough, but that maybe only because of my limited understanding of the data

hesa replied:
Using both. For spec, see https://spdx.github.io/spdx-spec/v2-draft/relationships-between-SPDX-elements/

priv-kweihmann commented on flict/flictlib/project/reader.py:
```diff
@@ -170,6 +183,22 @@ def _read_spdx_2_2(self, only_packages=None):
+                        else:
+                            _pkg = {
+                                'id': dep_id,
+                                'name': pkg.get('name', ""),
```
Again here the use of `get` might lead to more problems than it solves

hesa replied:
fixed in coming push

priv-kweihmann commented on flict/flictlib/project/reader.py:
```diff
@@ -200,9 +229,9 @@ def _flatten_package_tree(self, packages):
         for _package in packages.values():
             package_dict.update(
-                {f"{_package['name']}--{_package['version']}": _package},
+                {f"{_package['name']}--{_package.get('version', '')}": _package},
```
As `version` is mandatory (IMHO) we should keep the old implementation

hesa replied:
https://spdx.github.io/spdx-spec/package-information/#73-package-version-field
* 7.3 Package version field, Required | No

hesa replied:
But (woops) the above is the internal format and version should be present (could be '').
Using `_package['version']`

hesa committed:
Project: don't use .get, better to crash early

hesa committed:
REUSE: more detailed file (c) and lic info

hesa committed:
Test: new JSON data

hesa committed:
Example: Fix license inconsistency

hesa committed:
Project: Fix PR comments

Revert to old dependencies_license
Quotes: " -> '
Safer identification of spdxElementId

hesa committed:
Test: inner packages in SPDX

hesa committed:
Project: fix lint errors

hesa committed:
Example: remove files

hesa committed:
REUSE: GPL3 for non spdx

hesa committed:
Example: license for spdx examples

hesa committed:
REUSE: not used
